src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,268 @@
+        return isConditionalExpression(expression) ? getConditionalInfo(expression, checker) : getBinaryInfo(expression, checker);
+    }
+
+    function getConditionalInfo(expression: ConditionalExpression, checker: TypeChecker): Info | undefined {
+        const condition = expression.condition;
+        const lastPropertyAccessChain = getLastPropertyAccessChain(expression.whenTrue);
+
+        if (!lastPropertyAccessChain || checker.isNullableType(checker.getTypeAtLocation(lastPropertyAccessChain))) return undefined;
+
+        if ((isPropertyAccessExpression(condition) || isIdentifier(condition))
+            && getMatchingSubexpression(condition, lastPropertyAccessChain.expression, checker)) {
+            return { lastPropertyAccessChain, occurrences:[condition], expression };
+        }
+        else if (isBinaryExpression(condition)) {
+            const occurrences = getOccurrencesInExpression(lastPropertyAccessChain.expression, condition, checker);
+            return occurrences ? { lastPropertyAccessChain, occurrences, expression } : undefined;
+        }
+    }
+
+    function getBinaryInfo(expression: BinaryExpression, checker: TypeChecker): Info | undefined {
+        if (expression.operatorToken.kind !== SyntaxKind.AmpersandAmpersandToken) return undefined;
+        const lastPropertyAccessChain = getLastPropertyAccessChain(expression.right);
+
+        if (!lastPropertyAccessChain) return undefined;
+
+        const occurrences = getOccurrencesInExpression(lastPropertyAccessChain.expression, expression.left, checker);
+        return occurrences ? { lastPropertyAccessChain, occurrences, expression } : undefined;
+    }
+
+    /**
+     * Gets a list of property accesses that appear in matchTo and occur in sequence in expression.
+     */
+    function getOccurrencesInExpression(matchTo: Expression, expression: Expression, checker: TypeChecker): (PropertyAccessExpression | Identifier)[] | undefined {
+        const occurrences: (PropertyAccessExpression | Identifier)[] = [];
+        while (isBinaryExpression(expression) && expression.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken) {
+            const match = getMatchingSubexpression(matchTo, expression.right, checker);
+            if (!match) {
+                break;
+            }
+            occurrences.push(match);
+            matchTo = match;
+            expression = expression.left;
+        }
+        const finalMatch = getMatchingSubexpression(matchTo, expression, checker);
+        if (finalMatch) {
+            occurrences.push(finalMatch);
+        }
+        return occurrences.length > 0 ? occurrences: undefined;
+    }
+
+    /**
+     * Checks that expression is a syntactic subexpression of matchTo.
+     */
+    function getMatchingSubexpression(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
+        if (isCallExpression(matchTo)) {

```

Comment:
Can you give an example when `getMatchingSubexpression` will return something from a call expression? Even if getLastPropertyAccessChain starts returning call expressions, I can't think of an example. At least, I think the only way you'd see a call here is doubled or otherwise nested calls:

`a.b()()` or `a.b().c`

Reply:
Those were the cases, but I have moved the check into `chainStartsWith`